pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/JTypeMirror.java

```diff
@@ -0,0 +1,449 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.types;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.annotation.Experimental;
+import net.sourceforge.pmd.lang.java.ast.TypeNode;
+import net.sourceforge.pmd.lang.java.symbols.JClassSymbol;
+import net.sourceforge.pmd.lang.java.symbols.JExecutableSymbol;
+import net.sourceforge.pmd.lang.java.symbols.JMethodSymbol;
+import net.sourceforge.pmd.lang.java.symbols.JTypeDeclSymbol;
+import net.sourceforge.pmd.lang.java.symbols.JTypeParameterSymbol;
+import net.sourceforge.pmd.lang.java.types.JPrimitiveType.PrimitiveTypeKind;
+import net.sourceforge.pmd.lang.java.types.TypeOps.Convertibility;
+import net.sourceforge.pmd.lang.java.types.internal.infer.InferenceVar;
+
+/**
+ * Type mirrors represent Java types. They are created by a {@link TypeSystem}
+ * from {@link JTypeDeclSymbol symbols}, a layer of abstraction above reflection
+ * classes.
+ *
+ * <p>Type mirrors can be obtained {@linkplain TypesFromReflection from reflected types},
```

Comment:
Actually I think TypesFromReflection doesn't use the `Class` instances directly. It will ask the TypeSystem to resolve the class name, and gives up if that class is not on the classpath of the typesystem.

This is a remnant of an early version of the framework, where it was still using the reflected symbol implementation and not ASM. So this was effectively depended on by the reflected symbol implementation to get the type of eg a method parameter, and was important.

I figured, if we keep it we can build complex types with eg `typeSystem.fromReflect(new TypeLiteral<List<String>>() { })` (which would produce the type for `List<String>`) without having to go through hoops with `parameterize` and such. But it's not a crucial piece of the framework and could be removed. I think, we should decide when we try to port rules, to see if it carries its weight.